File: WasteVisionWebFE/src/data/mappers/user-mapper.ts
import { User } from "@/core/domain/User";
import {
  UserDTO,
  CreateUserDTO,
  UpdateUserDTO,
  UserSearchParamsDTO,
} from "../dto/user-dto";

export class UserMapper {
  /**
   * Maps a UserDTO to a domain User
   */
  public static toDomain(dto: UserDTO): User {
    return {
      id: dto.id,
      email: dto.email,
      username: dto.username,
      active: dto.active,
      role: {
        id: dto.roleId,
        description: dto.roleName || "",
        active: true,
      },
    };
  }

  /**
   * Maps a domain User to a UserDTO
   */
  public static toDTO(domain: User): UserDTO {
    return {
      id: domain.id,
      email: domain.email,
      roleId: domain.role?.id || "",
      username: domain.username,
      active: domain.active,
      role: domain.role,
    };
  }

  /**
   * Maps an array of UserDTOs to domain Users
   */
  public static toDomainList(dtos: UserDTO[]): User[] {
    return dtos.map((dto) => this.toDomain(dto));
  }

  /**
   * Maps a domain User to a CreateUserDTO
   */
  public static toCreateDTO(domain: User): CreateUserDTO {
    return {
      email: domain.email,
      username: domain.username,
      roleId: domain.role?.id || "",
    };
  }

  /**
   * Maps a domain User to an UpdateUserDTO
   */
  public static toUpdateDTO(domain: User): UpdateUserDTO {
    return {
      email: domain.email,
      username: domain.username,
      roleId: domain.role?.id || "",
    };
  }
}

File: WasteVisionWebBE/Domain/Users/UserMapper.cs
using System.Collections.Generic;
using System.Linq;
using DDDSample1.Domain.Roles;
using DDDSample1.Domain.Shared;

namespace DDDSample1.Domain.Users
{
    public static class UserMapper
    {
        public static UserDto ToDto(User user)
        {
            if (user == null)
                return null;

            return new UserDto(
                user.Id.AsString(),
                user.Email.AsString(),
                user.Username.AsString(),
                user.RoleId
            );
        }

        public static UserApiDto ToApiDto(User user)
        {
            if (user == null)
                return null;

            return new UserApiDto(
                user.Id.AsString(),
                user.Email.AsString(),
                user.Username.AsString(),
                user.Role?.Description?.AsString(),
                user.Role?.Id?.AsString(),
                user.Active
            );
        }

        public static User ToDomain(CreatingUserDto dto)
        {
            if (dto == null)
                return null;

            return new User(
                new Email(dto.Email),
                new Username(dto.Username),
                new RoleId(dto.RoleId)
            );
        }

        public static List<UserDto> ToDtoList(List<User> userList)
        {
            return userList.Select(user => ToDto(user)).ToList();
        }

        public static List<UserApiDto> ToApiDtoList(List<User> userList)
        {
            return userList.Select(user => ToApiDto(user)).ToList();
        }
    }
}
